### PR TITLE
If the inner Instantiator is cached we should just cache the kayo and…

### PR DESCRIPTION
**Pull request by ianoc**
… KryoPool also

Some of the WS looks off here as the file seems to have a mix of tabs and spaces. I'll follow up this PR with one replacing all tabs with spaces.

### CODE REVIEW

File: chill-hadoop/src/main/java/com/twitter/chill/hadoop/KryoSerialization.java [hunk 1]
```diff
@@ -38,6 +38,27 @@
...
+     * so here we piggy back on that to avoid generating new kryo's or kryo pools
+     */
+    public static synchronized void resetOrUpdateFromCache(KryoSerialization instance, KryoInstantiator kryoInst){
+      if(kryoInst != cachedKryoInst) {
+        cachedTestKryo = kryoInst.newKryo();
+        cachedPool = KryoPool.withByteArrayOutputStream(MAX_CACHED_KRYO, kryoInst);
+        cachedKryoInst = kryoInst;
+      }
+      instance.kryoPool = cachedPool;
+      instance.testKryo = cachedTestKryo;
```
Comment (johnynek) [inline chill-hadoop/src/main/java/com/twitter/chill/hadoop/KryoSerialization.java]: I'm worried about the thread-safety of testKryo. We could be sharing it across threads, and there is no guarantee it is thread-safe.

Could we either borrow from the KryoPool to test or use a lock to guard access?

**Comment by ianoc** [comment]
Updated this one to use the kryo pool for the tests

**Comment by johnynek** [comment]
this looks good. Have you tried a scalding E2E run with this?

This is binary compatible, right? Can republish.

**Comment by ianoc** [comment]
Would need to back port -- chill's dependency on algebird makes it not binary compatible at twitter currently. And yes I think so, can do a mima to be sure before publish anyway

**Comment by johnynek** [comment]
@ianoc is it really not compatible? I think the dependency is pretty weak. Excluding algebird here should work. (Actually algebird should be provided).

**Comment by ianoc** [comment]
It should be... but finding/hitting these has been pretty time consuming. Backporting is faster/easier in the end. (Scala compiler can detect some class you refer to in a jar isn't valid against a transitive dependency anymore). So I've backported changes... will see if I can test them tonight

**Comment by ianoc** [comment]
Twitter internal e2e's passed on this change fyi

**Comment by johnynek** [comment]
Awesome. Merge here and in a chill08 branch?

**Pull request merged**